pyVHDLParser/Blocks/Reference/Context.py

```diff
@@ -27,139 +27,329 @@
+
+	@classmethod
+	def stateWhitespace1(cls, parserState: TokenToBlockParser):
 		token = parserState.Token
-		if isinstance(token, WhitespaceToken):
-			parserState.NewToken =    BoundaryToken(fromExistingToken=token)
-			parserState.NextState =   cls.stateWhitespace1
-			return
-		elif isinstance(token, (LinebreakToken, CommentToken)):
-			block =                   LinebreakBlock if isinstance(token, LinebreakToken) else CommentBlock
-			parserState.NewBlock =    cls(parserState.LastBlock, parserState.TokenMarker, endToken=token.PreviousToken, multiPart=True)
-			_ =                       block(parserState.NewBlock, token)
-			parserState.TokenMarker = None
-			parserState.NextState =   cls.stateWhitespace1
+
+		if _isNonCodeToken(token):
+			parserState.NextState = cls.stateContextName
 			return
 
 		raise BlockParserException("Expected whitespace after keyword CONTEXT.", token)
 
 	@classmethod
-	def stateWhitespace1(cls, parserState: TokenToBlockParser):
+	def stateContextName(cls, parserState: TokenToBlockParser):
 		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			return
+
+		if isinstance(token, (WordToken, ExtendedIdentifier)):
+			parserState.NextState =     cls.stateWhitespaceOrSemicolonOrDot
+			return
+
+		raise BlockParserException("Expected context name (identifier).", token)
+
+	@classmethod
+	def stateWhitespaceOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			parserState.NextState =        cls.stateIsOrSemicolonOrDot
+			return
+
+		if isinstance(token, CharacterToken) and token == ";":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			parserState.Pop()
+			return
+		elif isinstance(token, CharacterToken) and token == ".":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			return
+
+		raise BlockParserException("Expected whitespace after context name (identifier).", token)
+
+	@classmethod
+	def stateIsOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if parserState.HandleNonCodeTokens(cls):
+			return
+
+		if isinstance(token, WordToken) and (token == "is"):
+			parserState.ReparseFromTokenMarker(DeclarationStartBlock.stateFromStartBlock)
+			return
+		elif isinstance(token, CharacterToken) and token == ";":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			parserState.Pop()
+			return
+		elif isinstance(token, CharacterToken) and token == ".":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			return
+
+		raise BlockParserException("Expected context name (identifier).", token)
+
+
+#
+# Context references
+#
+@export
+class ReferenceStartBlock(Block):
+	@classmethod
+	def stateFromStartBlock(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if isinstance(token, ContextKeyword):
+			parserState.NextState =   cls.stateWhitespace

```

Comment:
Do not create the `ReferenceStartBlock` here, so the behaviour of `stateContextKeyword` can be the same whether it is called from here, or from within a `DeclarationBody`.